[Flaky test] StarTreeClusterIntegrationTest

```
[INFO] Running org.apache.pinot.integration.tests.StarTreeClusterIntegrationTest
18:48:37.290 ERROR [ZkAsyncCallbacks] [ZkClient-EventThread-25-localhost:2191] Interrupted waiting for success
java.lang.InterruptedException: null
Dec 08, 2022 6:48:45 PM org.glassfish.grizzly.http.server.NetworkListener start
	at java.lang.Object.wait(Native Method) ~[?:?]
INFO: Started listener bound to [0.0.0.0:18998]
Dec 08, 2022 6:48:45 PM org.glassfish.grizzly.http.server.HttpServer start
INFO: [HttpServer] Started.
Dec 08, 2022 6:48:46 PM org.glassfish.grizzly.http.server.NetworkListener start
INFO: Started listener bound to [0.0.0.0:18099]
Dec 08, 2022 6:48:46 PM org.glassfish.grizzly.http.server.HttpServer start
INFO: [HttpServer-1] Started.
Dec 08, 2022 6:48:48 PM org.glassfish.grizzly.http.server.NetworkListener start
INFO: Started listener bound to [0.0.0.0:8097]
Dec 08, 2022 6:48:48 PM org.glassfish.grizzly.http.server.HttpServer start
INFO: [HttpServer-2] Started.
Dec 08, 2022 6:48:49 PM org.glassfish.grizzly.http.server.NetworkListener start
INFO: Started listener bound to [0.0.0.0:8096]
Dec 08, 2022 6:48:49 PM org.glassfish.grizzly.http.server.HttpServer start
INFO: [HttpServer-3] Started.
	at java.lang.Object.wait(Object.java:328) ~[?:?]
	at org.apache.helix.zookeeper.zkclient.callback.ZkAsyncCallbacks$DefaultCallback.waitForSuccess(ZkAsyncCallbacks.java:248) [helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkClient.issueSync(ZkClient.java:1616) [helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkClient.access$300(ZkClient.java:85) [helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkClient$4.run(ZkClient.java:1642) [helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkEventThread.run(ZkEventThread.java:99) [helix-core-1.0.4.jar:1.0.4]
18:48:39.597 WARN [ServiceStartableUtils] [main] Failed to find cluster config for cluster: StarTreeClusterIntegrationTest, skipping applying cluster config
18:48:41.875 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(1ac22bf7_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:41.904 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(085cbd9d_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:41.922 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(65bcae0d_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:41.934 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(8db083f1_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:41.953 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(3bf90c1e_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:41.970 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(81c29fc5_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:42.442 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(368d660f_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:42.500 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(8dd2221d_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:42.536 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(1271da3c_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:42.570 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(d1399f63_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:42.587 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(2c07fb5e_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
18:48:42.602 ERROR [DelayedAutoRebalancer] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(b00dd6f5_DEFAULT)] No instances or active instances available for resource leadControllerResource, allInstances: [], liveInstances: [], activeInstances: []
linux
18:48:47.287 WARN [ZKMetadataProvider] [main] Path: /CONFIGS/TABLE does not exist
18:48:47.290 WARN [HelixHelper] [main] Idempotent or null ideal state update for resource brokerResource, skipping update.
18:48:47.768 WARN [Tracing] [main] Using default implementation of thread accountant, due to invalid thread accountant factory null provided.
18:48:48.972 WARN [ControllerLeaderLocator] [main] Already created
18:48:48.974 WARN [Tracing] [main] Using default implementation of thread accountant, due to invalid thread accountant factory null provided.
Dec 08, 2022 6:49:39 PM org.glassfish.grizzly.http.server.NetworkListener shutdownNow
18:49:39.350 WARN [SegmentDeletionManager] [grizzly-http-server-3] Failed to find local segment file for segment file:/tmp/test-controller-data-dir1670525315444/mytable/mytable_16160_16189_6+%25
INFO: Stopped listener bound to [0.0.0.0:8097]
18:49:39.350 WARN [SegmentDeletionManager] [grizzly-http-server-3] Failed to find local segment file for segment file:/tmp/test-controller-data-dir1670525315444/mytable/mytable_16251_16281_9+%25
Dec 08, 2022 6:49:40 PM org.glassfish.grizzly.http.server.NetworkListener shutdownNow
18:49:39.350 WARN [SegmentDeletionManager] [grizzly-http-server-3] Failed to find local segment file for segment file:/tmp/test-controller-data-dir1670525315444/mytable/mytable_16313_16342_11+%25
INFO: Stopped listener bound to [0.0.0.0:8096]
18:49:39.350 WARN [SegmentDeletionManager] [grizzly-http-server-3] Failed to find local segment file for segment file:/tmp/test-controller-data-dir1670525315444/mytable/mytable_16343_16373_0+%25
Dec 08, 2022 6:49:40 PM org.glassfish.grizzly.http.server.NetworkListener shutdownNow
18:49:39.351 WARN [SegmentDeletionManager] [grizzly-http-server-3] Failed to find local segment file for segment file:/tmp/test-controller-data-dir1670525315444/mytable/mytable_16405_16435_2+%25
INFO: Stopped listener bound to [0.0.0.0:18099]
18:49:39.585 ERROR [mytable_OFFLINE-TableDeletionMessageHandler] [HelixTaskExecutor-message_handle_thread_53] onError: INTERNAL, CANCEL
Dec 08, 2022 6:49:40 PM org.glassfish.grizzly.http.server.NetworkListener shutdownNow
java.lang.InterruptedException: sleep interrupted
INFO: Stopped listener bound to [0.0.0.0:18998]
	at java.lang.Thread.sleep(Native Method) ~[?:?]
	at org.apache.pinot.server.starter.helix.HelixInstanceDataManager.deleteTable(HelixInstanceDataManager.java:216) ~[classes/:?]
	at org.apache.pinot.server.starter.helix.SegmentMessageHandlerFactory$TableDeletionMessageHandler.handleMessage(SegmentMessageHandlerFactory.java:174) ~[classes/:?]
	at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:97) [helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:49) [helix-core-1.0.4.jar:1.0.4]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
18:49:39.715 ERROR [DataTableHandler] [nioEventLoopGroup-2-1] Channel for server: localhost_O is now inactive, marking server down
18:49:39.981 ERROR [mytable_OFFLINE-TableDeletionMessageHandler] [HelixTaskExecutor-message_handle_thread_54] onError: INTERNAL, CANCEL
java.lang.InterruptedException: sleep interrupted
	at java.lang.Thread.sleep(Native Method) ~[?:?]
	at org.apache.pinot.server.starter.helix.HelixInstanceDataManager.deleteTable(HelixInstanceDataManager.java:216) ~[classes/:?]
	at org.apache.pinot.server.starter.helix.SegmentMessageHandlerFactory$TableDeletionMessageHandler.handleMessage(SegmentMessageHandlerFactory.java:174) ~[classes/:?]
	at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:97) [helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:49) [helix-core-1.0.4.jar:1.0.4]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
18:49:40.107 ERROR [DataTableHandler] [nioEventLoopGroup-2-2] Channel for server: localhost_O is now inactive, marking server down
18:49:40.284 WARN [ClusterChangeMediator] [Cleanup thread for StarTreeClusterIntegrationTest-Broker_10.1.0.183_18099-SPECTATOR] ClusterChangeMediator already stopped, skipping enqueuing the IDEAL_STATE change
18:49:40.286 WARN [ClusterChangeMediator] [Cleanup thread for StarTreeClusterIntegrationTest-Broker_10.1.0.183_18099-SPECTATOR] ClusterChangeMediator already stopped, skipping enqueuing the EXTERNAL_VIEW change
18:49:40.289 WARN [ClusterChangeMediator] [Cleanup thread for StarTreeClusterIntegrationTest-Broker_10.1.0.183_18099-SPECTATOR] ClusterChangeMediator already stopped, skipping enqueuing the INSTANCE_CONFIG change
18:49:40.537 ERROR [GenericHelixController] [HelixController-pipeline-default-StarTreeClusterIntegrationTest-(479e6d23_DEFAULT)] Exception while executing DEFAULT pipeline for cluster StarTreeClusterIntegrationTest. Will not continue to next pipeline
org.apache.helix.zookeeper.zkclient.exception.ZkInterruptedException: java.lang.InterruptedException
	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:2002) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkClient.getChildren(ZkClient.java:1347) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkClient.getChildren(ZkClient.java:1340) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.manager.zk.ZkBaseDataAccessor.getChildNames(ZkBaseDataAccessor.java:693) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.manager.zk.ZKHelixDataAccessor.getChildNames(ZKHelixDataAccessor.java:397) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.common.caches.PropertyCache.genSelectiveUpdateInput(PropertyCache.java:133) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.common.caches.PropertyCache.doRefreshWithSelectiveUpdate(PropertyCache.java:174) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.common.caches.PropertyCache.refresh(PropertyCache.java:158) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.controller.dataproviders.BaseControllerDataProvider.refreshLiveInstances(BaseControllerDataProvider.java:309) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.controller.dataproviders.BaseControllerDataProvider.doRefresh(BaseControllerDataProvider.java:418) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.controller.dataproviders.ResourceControllerDataProvider.refresh(ResourceControllerDataProvider.java:143) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.controller.stages.ReadClusterDataStage.process(ReadClusterDataStage.java:64) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:75) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:903) [helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:132) [helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1554) [helix-core-1.0.4.jar:1.0.4]
Caused by: java.lang.InterruptedException
	at org.apache.helix.zookeeper.zkclient.ZkConnection.handleInvokedMethodException(ZkConnection.java:290) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkConnection.getChildren(ZkConnection.java:183) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkClient$1.call(ZkClient.java:1353) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkClient$1.call(ZkClient.java:1347) ~[helix-core-1.0.4.jar:1.0.4]
	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1986) ~[helix-core-1.0.4.jar:1.0.4]
	... 15 more
Error:  Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 66.898 s <<< FAILURE! - in org.apache.pinot.integration.tests.StarTreeClusterIntegrationTest
Error:  org.apache.pinot.integration.tests.StarTreeClusterIntegrationTest.testGeneratedQueries  Time elapsed: 1.851 s  <<< FAILURE!
java.lang.AssertionError: 
Query comparison failed for: 
Star Query: SELECT Carrier, AirlineID, Flights, sum(ArrDel15), min(DepDel15), min(ArrDel15) FROM mytable WHERE Dest BETWEEN 'GRB' AND 'SJT' AND Flights IN (1) AND AirlineID <= 19690 GROUP BY Carrier, AirlineID, Flights ORDER BY Carrier, AirlineID, Flights
Star Response: {"resultTable":{"dataSchema":{"columnNames":["Carrier","AirlineID","Flights","sum(ArrDel15)","min(DepDel15)","min(ArrDel15)"],"columnDataTypes":["STRING","LONG","INT","DOUBLE","DOUBLE","DOUBLE"]},"rows":[["HA",19690,1,-749837.0,0.0,-9999.0],["WN",19393,1,-1.840513E7,-9999.0,-9999.0]]},"exceptions":[],"numServersQueried":2,"numServersResponded":2,"numSegmentsQueried":12,"numSegmentsProcessed":12,"numSegmentsMatched":12,"numConsumingSegmentsQueried":0,"numConsumingSegmentsProcessed":0,"numConsumingSegmentsMatched":0,"numDocsScanned":10420,"numEntriesScannedInFilter":59517,"numEntriesScannedPostFilter":62520,"numGroupsLimitReached":false,"totalDocs":115545,"timeUsedMs":18,"offlineThreadCpuTimeNs":10088600,"realtimeThreadCpuTimeNs":0,"offlineSystemActivitiesCpuTimeNs":23492305,"realtimeSystemActivitiesCpuTimeNs":0,"offlineResponseSerializationCpuTimeNs":37100,"realtimeResponseSerializationCpuTimeNs":0,"offlineTotalCpuTimeNs":33618005,"realtimeTotalCpuTimeNs":0,"segmentStatistics":[],"traceInfo":{},"minConsumingFreshnessTimeMs":0,"numRowsResultSet":2,"numSegmentsPrunedByBroker":0,"numSegmentsPrunedByServer":0,"numSegmentsPrunedInvalid":0,"numSegmentsPrunedByLimit":0,"numSegmentsPrunedByValue":0,"explainPlanNumEmptyFilterSegments":0,"explainPlanNumMatchAllFilterSegments":0}
Reference Query: SET useStarTree = false; SELECT Carrier, AirlineID, Flights, sum(ArrDel15), min(DepDel15), min(ArrDel15) FROM mytable WHERE Dest BETWEEN 'GRB' AND 'SJT' AND Flights IN (1) AND AirlineID <= 19690 GROUP BY Carrier, AirlineID, Flights ORDER BY Carrier, AirlineID, Flights
Reference Response: {"resultTable":{"dataSchema":{"columnNames":["Carrier","AirlineID","Flights","sum(ArrDel15)","min(DepDel15)","min(ArrDel15)"],"columnDataTypes":["STRING","LONG","INT","DOUBLE","DOUBLE","DOUBLE"]},"rows":[["HA",19690,1,-809830.0,0.0,-9999.0],["WN",19393,1,-1.840513E7,-9999.0,-9999.0]]},"exceptions":[],"numServersQueried":2,"numServersResponded":2,"numSegmentsQueried":12,"numSegmentsProcessed":12,"numSegmentsMatched":12,"numConsumingSegmentsQueried":0,"numConsumingSegmentsProcessed":0,"numConsumingSegmentsMatched":0,"numDocsScanned":14727,"numEntriesScannedInFilter":131387,"numEntriesScannedPostFilter":73635,"numGroupsLimitReached":false,"totalDocs":115545,"timeUsedMs":16,"offlineThreadCpuTimeNs":11989000,"realtimeThreadCpuTimeNs":0,"offlineSystemActivitiesCpuTimeNs":12899745,"realtimeSystemActivitiesCpuTimeNs":0,"offlineResponseSerializationCpuTimeNs":32800,"realtimeResponseSerializationCpuTimeNs":0,"offlineTotalCpuTimeNs":24921545,"realtimeTotalCpuTimeNs":0,"segmentStatistics":[],"traceInfo":{},"minConsumingFreshnessTimeMs":0,"numRowsResultSet":2,"numSegmentsPrunedByBroker":0,"numSegmentsPrunedByServer":0,"numSegmentsPrunedInvalid":0,"numSegmentsPrunedByLimit":0,"numSegmentsPrunedByValue":0,"explainPlanNumEmptyFilterSegments":0,"explainPlanNumMatchAllFilterSegments":0}: Iterators differ at element [1]: [["HA",19690,1,-809830.0,0.0,-9999.0],["WN",19393,1,-1.840513E7,-9999.0,-9999.0]] != [["HA",19690,1,-749837.0,0.0,-9999.0],["WN",19393,1,-1.840513E7,-9999.0,-9999.0]] expected [[["HA",19690,1,-809830.0,0.0,-9999.0],["WN",19393,1,-1.840513E7,-9999.0,-9999.0]]] but found [[["HA",19690,1,-749837.0,0.0,-9999.0],["WN",19393,1,-1.840513E7,-9999.0,-9999.0]]]
	at org.testng.Assert.fail(Assert.java:93)
	at org.testng.Assert.failNotEquals(Assert.java:512)
	at org.testng.Assert.assertEqualsImpl(Assert.java:134)
	at org.testng.Assert.assertEquals(Assert.java:610)
	at org.testng.Assert.assertEquals(Assert.java:661)
	at org.apache.pinot.integration.tests.StarTreeClusterIntegrationTest.testStarQuery(StarTreeClusterIntegrationTest.java:188)
	at org.apache.pinot.integration.tests.StarTreeClusterIntegrationTest.testGeneratedQueries(StarTreeClusterIntegrationTest.java:158)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:661)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:869)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1193)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:126)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:744)
	at org.testng.TestRunner.run(TestRunner.java:602)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
	at org.testng.TestNG.runSuites(TestNG.java:1144)
	at org.testng.TestNG.run(TestNG.java:1115)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:136)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:99)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:145)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)

Error:  org.apache.pinot.integration.tests.StarTreeClusterIntegrationTest.testPredicateOnMetrics  Time elapsed: 0.189 s
```

--- a/pinot-core/src/main/java/org/apache/pinot/core/startree/operator/StarTreeFilterOperator.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/startree/operator/StarTreeFilterOperator.java
@@ -203,12 +203,16 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
   @Nullable
   private StarTreeResult traverseStarTree() {
     MutableRoaringBitmap matchingDocIds = new MutableRoaringBitmap();
-    Set<String> globalRemainingPredicateColumns = Collections.emptySet();
-    boolean globalRemainingPredicateColumnsSet = false;
+    Set<String> globalRemainingPredicateColumns = null;
 
     StarTree starTree = _starTreeV2.getStarTree();
     List<String> dimensionNames = starTree.getDimensionNames();
     StarTreeNode starTreeRootNode = starTree.getRoot();
+
+    // Track whether we have found a leaf node added to the queue. If we have found a leaf node, and traversed to the
+    // level of the leave node, we can set globalRemainingPredicateColumns if not already set because we know the leaf
+    // node won't split further on other predicate columns.
+    boolean findLeafNode = starTreeRootNode.isLeaf();
 
     // Use BFS to traverse the star tree
     Queue<StarTreeNode> queue = new ArrayDeque<>();
@@ -216,6 +220,9 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
     int currentDimensionId = -1;
     Set<String> remainingPredicateColumns = new HashSet<>(_predicateEvaluatorsMap.keySet());
     Set<String> remainingGroupByColumns = new HashSet<>(_groupByColumns);
+    if (findLeafNode) {
+      globalRemainingPredicateColumns = new HashSet<>(remainingPredicateColumns);
+    }
     IntSet matchingDictIds = null;
     StarTreeNode starTreeNode;
     while ((starTreeNode = queue.poll()) != null) {
@@ -225,6 +232,9 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
         String dimension = dimensionNames.get(dimensionId);
         remainingPredicateColumns.remove(dimension);
         remainingGroupByColumns.remove(dimension);
+        if (findLeafNode && globalRemainingPredicateColumns == null) {
+          globalRemainingPredicateColumns = new HashSet<>(remainingPredicateColumns);
+        }
         matchingDictIds = null;
         currentDimensionId = dimensionId;
       }
@@ -240,14 +250,6 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
       // remaining predicate columns for this node
       if (starTreeNode.isLeaf()) {
         matchingDocIds.add((long) starTreeNode.getStartDocId(), starTreeNode.getEndDocId());
-        // Only set the global remaining predicate columns once because we traverse the tree with BFS, so the first leaf
-        // node always have all the remaining predicate columns
-        if (!globalRemainingPredicateColumnsSet) {
-          if (!remainingPredicateColumns.isEmpty()) {
-            globalRemainingPredicateColumns = new HashSet<>(remainingPredicateColumns);
-          }
-          globalRemainingPredicateColumnsSet = true;
-        }
         continue;
       }
 
@@ -257,8 +259,8 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
       // Only read star-node when the dimension is not in the global remaining predicate columns or group-by columns
       // because we cannot use star-node in such cases
       StarTreeNode starNode = null;
-      if (!globalRemainingPredicateColumns.contains(childDimension) && !remainingGroupByColumns.contains(
-          childDimension)) {
+      if ((globalRemainingPredicateColumns == null || !globalRemainingPredicateColumns.contains(childDimension))
+          && !remainingGroupByColumns.contains(childDimension)) {
         starNode = starTreeNode.getChildForDimensionValue(StarTreeNode.ALL);
       }
 
@@ -287,18 +289,22 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
           // star-node if so
           if (starNode != null && numMatchingDictIds >= numChildren - 1) {
             List<StarTreeNode> matchingChildNodes = new ArrayList<>();
+            boolean findLeafChildNode = false;
             while (childrenIterator.hasNext()) {
               StarTreeNode childNode = childrenIterator.next();
               if (matchingDictIds.contains(childNode.getDimensionValue())) {
                 matchingChildNodes.add(childNode);
+                findLeafChildNode |= childNode.isLeaf();
               }
             }
             if (matchingChildNodes.size() == numChildren - 1) {
               // All the child nodes (except for the star-node) match the predicate, use the star-node
               queue.add(starNode);
+              findLeafNode |= starNode.isLeaf();
             } else {
               // Some child nodes do not match the predicate, use the matching child nodes
               queue.addAll(matchingChildNodes);
+              findLeafNode |= findLeafChildNode;
             }
           } else {
             // Cannot use the star-node, use the matching child nodes
@@ -306,6 +312,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
               StarTreeNode childNode = childrenIterator.next();
               if (matchingDictIds.contains(childNode.getDimensionValue())) {
                 queue.add(childNode);
+                findLeafNode |= childNode.isLeaf();
               }
             }
           }
@@ -318,6 +325,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
             // Child node might be null because the matching dictionary id might not exist under this branch
             if (childNode != null) {
               queue.add(childNode);
+              findLeafNode |= childNode.isLeaf();
             }
           }
         }
@@ -327,6 +335,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
         if (starNode != null) {
           // Star-node exists, use it
           queue.add(starNode);
+          findLeafNode |= starNode.isLeaf();
         } else {
           // Star-node does not exist or cannot be used, add all non-star nodes to the queue
           Iterator<? extends StarTreeNode> childrenIterator = starTreeNode.getChildrenIterator();
@@ -334,13 +343,15 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
             StarTreeNode childNode = childrenIterator.next();
             if (childNode.getDimensionValue() != StarTreeNode.ALL) {
               queue.add(childNode);
+              findLeafNode |= childNode.isLeaf();
             }
           }
         }
       }
     }
 
-    return new StarTreeResult(matchingDocIds, globalRemainingPredicateColumns);
+    return new StarTreeResult(matchingDocIds,
+        globalRemainingPredicateColumns != null ? globalRemainingPredicateColumns : Collections.emptySet());
   }
 
   /**

--- a/pinot-core/src/main/java/org/apache/pinot/core/startree/operator/StarTreeFilterOperator.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/startree/operator/StarTreeFilterOperator.java
@@ -212,7 +212,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
     // Track whether we have found a leaf node added to the queue. If we have found a leaf node, and traversed to the
     // level of the leave node, we can set globalRemainingPredicateColumns if not already set because we know the leaf
     // node won't split further on other predicate columns.
-    boolean findLeafNode = starTreeRootNode.isLeaf();
+    boolean foundLeafNode = starTreeRootNode.isLeaf();
 
     // Use BFS to traverse the star tree
     Queue<StarTreeNode> queue = new ArrayDeque<>();
@@ -220,7 +220,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
     int currentDimensionId = -1;
     Set<String> remainingPredicateColumns = new HashSet<>(_predicateEvaluatorsMap.keySet());
     Set<String> remainingGroupByColumns = new HashSet<>(_groupByColumns);
-    if (findLeafNode) {
+    if (foundLeafNode) {
       globalRemainingPredicateColumns = new HashSet<>(remainingPredicateColumns);
     }
     IntSet matchingDictIds = null;
@@ -232,7 +232,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
         String dimension = dimensionNames.get(dimensionId);
         remainingPredicateColumns.remove(dimension);
         remainingGroupByColumns.remove(dimension);
-        if (findLeafNode && globalRemainingPredicateColumns == null) {
+        if (foundLeafNode && globalRemainingPredicateColumns == null) {
           globalRemainingPredicateColumns = new HashSet<>(remainingPredicateColumns);
         }
         matchingDictIds = null;
@@ -300,11 +300,11 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
             if (matchingChildNodes.size() == numChildren - 1) {
               // All the child nodes (except for the star-node) match the predicate, use the star-node
               queue.add(starNode);
-              findLeafNode |= starNode.isLeaf();
+              foundLeafNode |= starNode.isLeaf();
             } else {
               // Some child nodes do not match the predicate, use the matching child nodes
               queue.addAll(matchingChildNodes);
-              findLeafNode |= findLeafChildNode;
+              foundLeafNode |= findLeafChildNode;
             }
           } else {
             // Cannot use the star-node, use the matching child nodes
@@ -312,7 +312,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
               StarTreeNode childNode = childrenIterator.next();
               if (matchingDictIds.contains(childNode.getDimensionValue())) {
                 queue.add(childNode);
-                findLeafNode |= childNode.isLeaf();
+                foundLeafNode |= childNode.isLeaf();
               }
             }
           }
@@ -325,7 +325,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
             // Child node might be null because the matching dictionary id might not exist under this branch
             if (childNode != null) {
               queue.add(childNode);
-              findLeafNode |= childNode.isLeaf();
+              foundLeafNode |= childNode.isLeaf();
             }
           }
         }
@@ -335,7 +335,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
         if (starNode != null) {
           // Star-node exists, use it
           queue.add(starNode);
-          findLeafNode |= starNode.isLeaf();
+          foundLeafNode |= starNode.isLeaf();
         } else {
           // Star-node does not exist or cannot be used, add all non-star nodes to the queue
           Iterator<? extends StarTreeNode> childrenIterator = starTreeNode.getChildrenIterator();
@@ -343,7 +343,7 @@ public class StarTreeFilterOperator extends BaseFilterOperator {
             StarTreeNode childNode = childrenIterator.next();
             if (childNode.getDimensionValue() != StarTreeNode.ALL) {
               queue.add(childNode);
-              findLeafNode |= childNode.isLeaf();
+              foundLeafNode |= childNode.isLeaf();
             }
           }
         }

--- a/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/StarTreeClusterIntegrationTest.java
+++ b/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/StarTreeClusterIntegrationTest.java
@@ -24,6 +24,7 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Random;
 import org.apache.commons.io.FileUtils;
 import org.apache.pinot.integration.tests.startree.SegmentInfoProvider;
 import org.apache.pinot.integration.tests.startree.StarTreeQueryGenerator;
@@ -33,10 +34,11 @@ import org.apache.pinot.spi.config.table.StarTreeIndexConfig;
 import org.apache.pinot.spi.config.table.TableConfig;
 import org.apache.pinot.spi.data.Schema;
 import org.apache.pinot.util.TestUtils;
-import org.testng.Assert;
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
 
 
 /**
@@ -65,6 +67,9 @@ public class StarTreeClusterIntegrationTest extends BaseClusterIntegrationTest {
           AggregationFunctionType.SUM, AggregationFunctionType.AVG, AggregationFunctionType.MINMAXRANGE,
           AggregationFunctionType.DISTINCTCOUNTBITMAP);
   private static final int NUM_QUERIES_TO_GENERATE = 100;
+
+  private final long _randomSeed = System.currentTimeMillis();
+  private final Random _random = new Random(_randomSeed);
 
   private StarTreeQueryGenerator _starTree1QueryGenerator;
   private StarTreeQueryGenerator _starTree2QueryGenerator;
@@ -95,27 +100,26 @@ public class StarTreeClusterIntegrationTest extends BaseClusterIntegrationTest {
     Schema schema = createSchema();
     addSchema(schema);
 
-    // Randomly pick some dimensions and metrics for star-trees
-    List<String> starTree1Dimensions = new ArrayList<>(NUM_STAR_TREE_DIMENSIONS);
-    List<String> starTree2Dimensions = new ArrayList<>(NUM_STAR_TREE_DIMENSIONS);
+    // Pick fixed dimensions and metrics for the first star-tree
+    List<String> starTree1Dimensions =
+        Arrays.asList("OriginCityName", "DepTimeBlk", "LongestAddGTime", "CRSDepTime", "DivArrDelay");
+    List<String> starTree1Metrics =
+        Arrays.asList("CarrierDelay", "DepDelay", "LateAircraftDelay", "ArrivalDelayGroups", "ArrDel15");
+    int starTree1MaxLeafRecords = 10;
+
+    // Randomly pick some dimensions and metrics for the second star-tree
     List<String> allDimensions = new ArrayList<>(schema.getDimensionNames());
-    Collections.shuffle(allDimensions);
-    for (int i = 0; i < NUM_STAR_TREE_DIMENSIONS; i++) {
-      starTree1Dimensions.add(allDimensions.get(2 * i));
-      starTree2Dimensions.add(allDimensions.get(2 * i + 1));
-    }
-    List<String> starTree1Metrics = new ArrayList<>(NUM_STAR_TREE_METRICS);
-    List<String> starTree2Metrics = new ArrayList<>(NUM_STAR_TREE_METRICS);
+    Collections.shuffle(allDimensions, _random);
+    List<String> starTree2Dimensions = allDimensions.subList(0, NUM_STAR_TREE_DIMENSIONS);
     List<String> allMetrics = new ArrayList<>(schema.getMetricNames());
-    Collections.shuffle(allMetrics);
-    for (int i = 0; i < NUM_STAR_TREE_METRICS; i++) {
-      starTree1Metrics.add(allMetrics.get(2 * i));
-      starTree2Metrics.add(allMetrics.get(2 * i + 1));
-    }
+    Collections.shuffle(allMetrics, _random);
+    List<String> starTree2Metrics = allMetrics.subList(0, NUM_STAR_TREE_METRICS);
+    int starTree2MaxLeafRecords = 100;
+
     TableConfig tableConfig = createOfflineTableConfig();
     tableConfig.getIndexingConfig().setStarTreeIndexConfigs(
-        Arrays.asList(getStarTreeIndexConfig(starTree1Dimensions, starTree1Metrics),
-            getStarTreeIndexConfig(starTree2Dimensions, starTree2Metrics)));
+        Arrays.asList(getStarTreeIndexConfig(starTree1Dimensions, starTree1Metrics, starTree1MaxLeafRecords),
+            getStarTreeIndexConfig(starTree2Dimensions, starTree2Metrics, starTree2MaxLeafRecords)));
     addTableConfig(tableConfig);
 
     // Unpack the Avro files
@@ -132,22 +136,23 @@ public class StarTreeClusterIntegrationTest extends BaseClusterIntegrationTest {
       aggregationFunctions.add(functionType.getName());
     }
     _starTree1QueryGenerator = new StarTreeQueryGenerator(DEFAULT_TABLE_NAME, starTree1Dimensions, starTree1Metrics,
-        segmentInfoProvider.getSingleValueDimensionValuesMap(), aggregationFunctions);
+        segmentInfoProvider.getSingleValueDimensionValuesMap(), aggregationFunctions, _random);
     _starTree2QueryGenerator = new StarTreeQueryGenerator(DEFAULT_TABLE_NAME, starTree2Dimensions, starTree2Metrics,
-        segmentInfoProvider.getSingleValueDimensionValuesMap(), aggregationFunctions);
+        segmentInfoProvider.getSingleValueDimensionValuesMap(), aggregationFunctions, _random);
 
     // Wait for all documents loaded
     waitForAllDocsLoaded(600_000L);
   }
 
-  private static StarTreeIndexConfig getStarTreeIndexConfig(List<String> dimensions, List<String> metrics) {
+  private static StarTreeIndexConfig getStarTreeIndexConfig(List<String> dimensions, List<String> metrics,
+      int maxLeafRecords) {
     List<String> functionColumnPairs = new ArrayList<>();
     for (AggregationFunctionType functionType : AGGREGATION_FUNCTION_TYPES) {
       for (String metric : metrics) {
         functionColumnPairs.add(new AggregationFunctionColumnPair(functionType, metric).toColumnName());
       }
     }
-    return new StarTreeIndexConfig(dimensions, null, functionColumnPairs, 100);
+    return new StarTreeIndexConfig(dimensions, null, functionColumnPairs, maxLeafRecords);
   }
 
   @Test
@@ -160,23 +165,13 @@ public class StarTreeClusterIntegrationTest extends BaseClusterIntegrationTest {
   }
 
   @Test
-  public void testPredicateOnMetrics()
+  public void testHardCodedQueries()
       throws Exception {
-    String starQuery;
-
-    // Query containing predicate on one metric only
-    starQuery = "SELECT SUM(DepDelayMinutes) FROM mytable WHERE DepDelay > 0";
-    testStarQuery(starQuery);
-    starQuery = "SELECT SUM(DepDelayMinutes) FROM mytable WHERE DepDelay BETWEEN 0 and 10000";
-    testStarQuery(starQuery);
-
-    // Query containing predicate on multiple metrics
-    starQuery = "SELECT SUM(DepDelayMinutes) FROM mytable WHERE DepDelay > 0 AND ArrDelay > 0";
-    testStarQuery(starQuery);
-
-    // Query containing predicate on multiple metrics and dimensions
-    starQuery = "SELECT SUM(DepDelayMinutes) FROM mytable WHERE DepDelay > 0 AND ArrDelay > 0 AND OriginStateName = "
-        + "'Massachusetts'";
+    // This query can test the case of one predicate matches all the child nodes but star-node cannot be used because
+    // the predicate is included as remaining predicate from another branch
+    String starQuery = "SELECT DepTimeBlk, COUNT(*) FROM mytable "
+        + "WHERE CRSDepTime BETWEEN 1137 AND 1849 AND DivArrDelay > 218 AND CRSDepTime NOT IN (35, 1633, 1457, 140) "
+        + "AND LongestAddGTime NOT IN (17, 105, 20, 22) GROUP BY DepTimeBlk ORDER BY DepTimeBlk";
     testStarQuery(starQuery);
   }
 
@@ -185,9 +180,10 @@ public class StarTreeClusterIntegrationTest extends BaseClusterIntegrationTest {
     String referenceQuery = "SET useStarTree = false; " + starQuery;
     JsonNode starResponse = postQuery(starQuery);
     JsonNode referenceResponse = postQuery(referenceQuery);
-    Assert.assertEquals(starResponse.get("resultTable"), referenceResponse.get("resultTable"),
-        "Query comparison failed for: \nStar Query: " + starQuery + "\nStar Response: " + starResponse
-            + "\nReference Query: " + referenceQuery + "\nReference Response: " + referenceResponse);
+    assertEquals(starResponse.get("resultTable"), referenceResponse.get("resultTable"), String.format(
+        "Query comparison failed for: \n"
+            + "Star Query: %s\nStar Response: %s\nReference Query: %s\nReference Response: %s\nRandom Seed: %d",
+        starQuery, starResponse, referenceQuery, referenceResponse, _randomSeed));
   }
 
   @AfterClass

--- a/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/startree/StarTreeQueryGenerator.java
+++ b/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/startree/StarTreeQueryGenerator.java
@@ -51,7 +51,6 @@ public class StarTreeQueryGenerator {
   private static final int MAX_NUM_GROUP_BYS = 3;
   private static final int MAX_NUM_IN_VALUES = 5;
   private static final int SHUFFLE_THRESHOLD = 5 * MAX_NUM_IN_VALUES;
-  private static final Random RANDOM = new Random();
   // Add more comparators here to generate them in the 'WHERE' clause.
   private static final List<String> COMPARATORS = Arrays.asList("=", "<>", "<", ">", "<=", ">=");
 
@@ -60,14 +59,16 @@ public class StarTreeQueryGenerator {
   private final List<String> _metricColumns;
   private final Map<String, List<Object>> _singleValueDimensionValuesMap;
   private final List<String> _aggregationFunctions;
+  private final Random _random;
 
   public StarTreeQueryGenerator(String tableName, List<String> singleValueDimensionColumns, List<String> metricColumns,
-      Map<String, List<Object>> singleValueDimensionValuesMap, List<String> aggregationFunctions) {
+      Map<String, List<Object>> singleValueDimensionValuesMap, List<String> aggregationFunctions, Random random) {
     _tableName = tableName;
     _singleValueDimensionColumns = singleValueDimensionColumns;
     _metricColumns = metricColumns;
     _singleValueDimensionValuesMap = singleValueDimensionValuesMap;
     _aggregationFunctions = aggregationFunctions;
+    _random = random;
   }
 
   /**
@@ -77,7 +78,7 @@ public class StarTreeQueryGenerator {
    * @return aggregation function.
    */
   private String generateAggregation(String metricColumn) {
-    return String.format("%s(%s)", _aggregationFunctions.get(RANDOM.nextInt(_aggregationFunctions.size())),
+    return String.format("%s(%s)", _aggregationFunctions.get(_random.nextInt(_aggregationFunctions.size())),
         metricColumn);
   }
 
@@ -87,11 +88,11 @@ public class StarTreeQueryGenerator {
    * @return aggregation section.
    */
   private String generateAggregations() {
-    int numAggregations = RANDOM.nextInt(MAX_NUM_AGGREGATIONS) + 1;
+    int numAggregations = _random.nextInt(MAX_NUM_AGGREGATIONS) + 1;
     int numMetrics = _metricColumns.size();
     String[] aggregations = new String[numAggregations];
     for (int i = 0; i < numAggregations; i++) {
-      aggregations[i] = generateAggregation(_metricColumns.get(RANDOM.nextInt(numMetrics)));
+      aggregations[i] = generateAggregation(_metricColumns.get(_random.nextInt(numMetrics)));
     }
     return StringUtils.join(aggregations, ", ");
   }
@@ -105,10 +106,10 @@ public class StarTreeQueryGenerator {
   private String generateComparisonPredicate(String dimensionColumn) {
     StringBuilder stringBuilder = new StringBuilder(dimensionColumn);
 
-    stringBuilder.append(' ').append(COMPARATORS.get(RANDOM.nextInt(COMPARATORS.size()))).append(' ');
+    stringBuilder.append(' ').append(COMPARATORS.get(_random.nextInt(COMPARATORS.size()))).append(' ');
 
     List<Object> valueArray = _singleValueDimensionValuesMap.get(dimensionColumn);
-    Object value = valueArray.get(RANDOM.nextInt(valueArray.size()));
+    Object value = valueArray.get(_random.nextInt(valueArray.size()));
     if (value instanceof String) {
       stringBuilder.append('\'').append(((String) value).replaceAll("'", "''")).append('\'');
     } else {
@@ -128,8 +129,8 @@ public class StarTreeQueryGenerator {
     StringBuilder stringBuilder = new StringBuilder(dimensionColumn).append(BETWEEN);
 
     List<Object> valueArray = _singleValueDimensionValuesMap.get(dimensionColumn);
-    Object value1 = valueArray.get(RANDOM.nextInt(valueArray.size()));
-    Object value2 = valueArray.get(RANDOM.nextInt(valueArray.size()));
+    Object value1 = valueArray.get(_random.nextInt(valueArray.size()));
+    Object value2 = valueArray.get(_random.nextInt(valueArray.size()));
 
     Preconditions.checkState((value1 instanceof String && value2 instanceof String) || (value1 instanceof Number
         && value2 instanceof Number));
@@ -163,7 +164,7 @@ public class StarTreeQueryGenerator {
    */
   private String generateInPredicate(String dimensionColumn) {
     StringBuilder stringBuilder = new StringBuilder(dimensionColumn);
-    if (RANDOM.nextBoolean()) {
+    if (_random.nextBoolean()) {
       stringBuilder.append(IN).append('(');
     } else {
       stringBuilder.append(NOT_IN).append('(');
@@ -171,7 +172,7 @@ public class StarTreeQueryGenerator {
 
     List<Object> valueArray = _singleValueDimensionValuesMap.get(dimensionColumn);
     int size = valueArray.size();
-    int numValues = Math.min(RANDOM.nextInt(MAX_NUM_IN_VALUES) + 1, size);
+    int numValues = Math.min(_random.nextInt(MAX_NUM_IN_VALUES) + 1, size);
     if (size < SHUFFLE_THRESHOLD) {
       // For smaller size values, use shuffle strategy.
       Collections.shuffle(valueArray);
@@ -190,7 +191,7 @@ public class StarTreeQueryGenerator {
       // For larger size values, use random indices strategy.
       Set<Integer> indices = new HashSet<>();
       while (indices.size() < numValues) {
-        indices.add(RANDOM.nextInt(size));
+        indices.add(_random.nextInt(size));
       }
       boolean isFirst = true;
       for (int index : indices) {
@@ -218,7 +219,7 @@ public class StarTreeQueryGenerator {
    */
   @Nullable
   private String generatePredicates() {
-    int numPredicates = RANDOM.nextInt(MAX_NUM_PREDICATES + 1);
+    int numPredicates = _random.nextInt(MAX_NUM_PREDICATES + 1);
     if (numPredicates == 0) {
       return null;
     }
@@ -230,8 +231,8 @@ public class StarTreeQueryGenerator {
       if (i != 0) {
         stringBuilder.append(AND);
       }
-      String dimensionName = _singleValueDimensionColumns.get(RANDOM.nextInt(numDimensions));
-      switch (RANDOM.nextInt(3)) {
+      String dimensionName = _singleValueDimensionColumns.get(_random.nextInt(numDimensions));
+      switch (_random.nextInt(3)) {
         case 0:
           stringBuilder.append(generateComparisonPredicate(dimensionName));
           break;
@@ -251,7 +252,7 @@ public class StarTreeQueryGenerator {
    * Randomly generate the group-by columns, may return {@code null}.
    */
   private String generateGroupByColumns() {
-    int numColumns = RANDOM.nextInt(MAX_NUM_GROUP_BYS + 1);
+    int numColumns = _random.nextInt(MAX_NUM_GROUP_BYS + 1);
     if (numColumns == 0) {
       return null;
     }

--- a/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/v2/builder/BaseSingleTreeBuilder.java
+++ b/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/v2/builder/BaseSingleTreeBuilder.java
@@ -380,12 +380,12 @@ abstract class BaseSingleTreeBuilder implements SingleTreeBuilder {
         nodeDimensionValue = dimensionValue;
       }
     }
-    TreeNode laseNode = getNewNode();
-    laseNode._dimensionId = dimensionId;
-    laseNode._dimensionValue = nodeDimensionValue;
-    laseNode._startDocId = nodeStartDocId;
-    laseNode._endDocId = endDocId;
-    nodes.put(nodeDimensionValue, laseNode);
+    TreeNode lastNode = getNewNode();
+    lastNode._dimensionId = dimensionId;
+    lastNode._dimensionValue = nodeDimensionValue;
+    lastNode._startDocId = nodeStartDocId;
+    lastNode._endDocId = endDocId;
+    nodes.put(nodeDimensionValue, lastNode);
     return nodes;
   }
 

